tensorflow/core/kernels/fused_batch_norm_op.h

```diff
@@ -43,6 +45,66 @@ struct InvVarianceToVariance {
                   int channels, T* variance);
 };
+
+
+// Functor used by FusedBatchNormGradOp to do the computations when is_training=False.
+// Both CPU and GPU will use this functor.
+template <typename Device, typename T>
+struct FusedBatchNormFreezeGrad {
+  void operator()(const Device& d, const Tensor& y_backprop_input,
+                  const Tensor& x_input, const Tensor& scale_input,
+                  const Tensor& mean_input, const Tensor& variance_input,
+                  T epsilon, Tensor* x_backprop_output,
+                  Tensor* scale_backprop_output, Tensor* offset_backprop_output,
+                  typename TTypes<T>::Vec scratch1, typename TTypes<T>::Vec scratch2) {
+    typename TTypes<T, 4>::ConstTensor out_backprop(y_backprop_input.tensor<T, 4>());
+    typename TTypes<T, 4>::ConstTensor input(x_input.tensor<T, 4>());
+    typename TTypes<T>::ConstVec scale(scale_input.vec<T>());
+    typename TTypes<T>::ConstVec mean(mean_input.vec<T>());
+    typename TTypes<T>::ConstVec var(variance_input.vec<T>());
+    typename TTypes<T, 4>::Tensor x_backprop(x_backprop_output->tensor<T, 4>());
+    typename TTypes<T>::Vec scale_backprop(scale_backprop_output->vec<T>());
+    typename TTypes<T>::Vec offset_backprop(offset_backprop_output->vec<T>());
+
+    const int depth = mean.dimension(0);
+    const int rest_size = input.size() / depth;
+
+    Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);
+#if !defined(EIGEN_HAS_INDEX_LIST)
+    Eigen::DSizes<Eigen::Index, 2> one_by_depth(1, depth);
+    Eigen::array<int, 1> reduction_axis{0};
+    Eigen::array<int, 2> rest_by_one({rest_size, 1});
+#else
+    Eigen::IndexList<Eigen::type2index<1>, Eigen::Index> one_by_depth;
+    one_by_depth.set(1, depth);
+    Eigen::IndexList<Eigen::type2index<0> > reduction_axis;
+    Eigen::IndexList<Eigen::Index, Eigen::type2index<1> > rest_by_one;
+    rest_by_one.set(0, rest_size);
+#endif
+
+    // db = out_backprop
+    // dg = out_backprop * ((x - m) * rsqrt(v + epsilon))
+    // dx = out_backprop * (gamma * rsqrt(v + epsilon))
+    offset_backprop.device(d) = out_backprop.reshape(rest_by_depth).sum(reduction_axis);
+
+    // scratch1 = rsqrt(v + epsilon)
+    scratch1.device(d) = (var + var.constant(epsilon)).rsqrt();
+
+    // scratch2 = sum_over_rest(out_backprop * (x - m))
+    scratch2.device(d) = (out_backprop.reshape(rest_by_depth) *
+                          (input.reshape(rest_by_depth) -
+                           mean.reshape(one_by_depth).broadcast(rest_by_one)))
+                           .sum(reduction_axis);
+
+    x_backprop.reshape(rest_by_depth).device(d) =
+        out_backprop.reshape(rest_by_depth) * ((scratch1 * scale)
+                                               .eval()
+                                               .reshape(one_by_depth)
+                                               .broadcast(rest_by_one));
+    scale_backprop.device(d) = scratch2 * scratch1;
```

Comment:
Are what implemented above equivalent to python implementation below? 
grad_offset = reduce_sum(grad_y)
grad_scale = reduce_sum(grad_y*(x-pop_mean)*var_rsqrt)
grad_x = grad_y * scale * var_rsqrt

Reply:
That looks equivalent to me